### PR TITLE
Implemented Required Features

### LINKED ISSUE #1: Image Histogram
## Summary of the request

Create a new handle that allows picking one image from a list and returns the histogram of an image. The histogram can be rendered either in the frontend (JavaScript library, for extra points), or in the backend and passed as an image (python / matplotlib).
**This is a new feature. Old functionalities must be preserved.**
Difficulty: 
image/image(If using Javascript)

## Subtasks
* Copy the classification API form and modify it where it is required.
* Implement the new API that takes as input the id of an image and returns the histogram.
* Visualize the histogram and the image in the GUI. 

## Plan for the implementation

Modify the code such that the new interface shows the dropdown menu to select the image and the submit button.
It is not required to use queues. The API can be implemented as showing the form if a GET request is issued, and the results if the POST is validated.

image

Useful links: 
* https://stackoverflow.com/questions/22159160/python-calculate-histogram-of-image

### LINKED ISSUE #2: Image transformation
## Summary of the request

Create a handler that allows picking an image and set parameters for an image transformation (https://pillow.readthedocs.io/en/stable/reference/ImageEnhance.html). The user should be allowed to specify the color, brightness, contrast, and sharpness to apply to the original image. The transformed image should appear in the output interface.
**This is a new feature. Old functionalities must be preserved.**
Difficulty:
image

## Subtasks

* Copy the classification API interface.
* Create a form for selecting the transformation.
* Implement the API that applies the transformation to the selected image.
* Implement the visualization of results.

## Plan for the implementation

Modify the code such that the new interface shows the dropdown menu to select the image, the submit button, and a NEW field that allows specifying values for the transformation to apply. The values for the transformation should have default values, but should be editable by the user. Use Flask forms to create the interface for selecting the transformation. Use correct field types for the inputs.
It is not required to use queues. The API can be implemented as showing the form if a GET request is issued, and the results if the POST is validated.

Useful links:
* https://pillow.readthedocs.io/en/stable/reference/ImageEnhance.html
* [Flask form fields](https://wtforms.readthedocs.io/en/2.3.x/fields/)

### LINKED ISSUE #3: Download results buttons
## Summary of the request

Add the “download results” button to the web interface. The button should allow the user to download the results (classification scores) as a JSON file and as a png file showing the top 5 scores in a plot (bar chart).
**All the functionalities must be preserved, this is just a new feature**.
Difficulty:
image

## Subtasks

* Add the "download results" button to the interface.
* Add the "download plot" button to the interface.
* Implement the API for returning a JSON file with the results.
* Implement the API for returning the plot.
* Connect the buttons with the APIs.

## Plan for the implementation

The easiest way to implement the new functionality is to create the files on the go when the button is pressed, through the button URL linked with an HTML tag (```<a href="URL">```, see references below).

image

Useful links:
* https://www.w3schools.com/tags/tag_a.asp
* https://stackoverflow.com/questions/30024948/flask-download-a-csv-file-on-clicking-a-button

### LINKED ISSUE #4: Upload Image Button
## Summary of the request

Add the “upload image” functionality. Implement a new route that instead of showing the dropdown list of images to select, allows the user to select an image from the computer, and returns the classification results.
The button should allow the user to upload a custom image and classify it. The image should be displayed in the output interface along with the classification scores.
**The old functionalities should be preserved, this is an additional feature**.
Difficulty:
image

## Subtasks

* Copy the route and HTML file of the classification API.
* Add the new route to the routes init file.
* Add "upload" button in the frontend (HTML).
* Implement the API for storing the image.
* Pass the image to the classification API.

## Plan for the implementation

image

Implement using the Flask APIs and modifying a copy of the available classification API. No particular additional library is required.

Useful links:
* https://flask.palletsprojects.com/en/2.0.x/patterns/fileuploads/

### PR DESCRIPTION
**Pull request by AntonioCampus**
## Changelog
* 

* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

- [ ] Image upload: this feature allows the user to classify a custom image.
- [ ] histogram: this feature allows to display the image histogram of an image
- [ ] download: the user can download the resulting of classification by downloading the plot and the json file
- [ ] image transformation: it allows to change Brightness, Contrast, Sharpness and color of an image

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:(does the pr fix some issues? Tag them with #<issue id>)

  Fixes 

 * #1  #2 #3  #4

### CODE REVIEW

**Comment by AntonioCampus** [comment]
Hi @maurapintor github doesn't allow me to add you as a reviewer

**Comment by maurapintor** [comment]
> Hi @maurapintor github doesn't allow me to add you as a reviewer

I know, it's very strange but happened to others as well. Thank you for letting me know. I will have a look at the PR as soon as possible!

File: app/routes/classifications_by_uploading.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        
+        if not allowed_file(file.filename):
+            return redirect(request.url)
+        
+        filename = secure_filename(file.filename)
+        absolute_path = os.path.join(app.config['UPLOAD_FOLDER'], filename)
+        print("apth : {}".format(absolute_path))
+        file.save(absolute_path)
+
+        # new part 23/11/22
```
Comment (maurapintor) [inline app/routes/classifications_by_uploading.py]: Are these comments needed? If not, I suggest removing them.

File: app/routes/classifications.py [hunk 2]
```diff
@@ -21,6 +21,12 @@ def classifications():
         image_id = form.image.data
         model_id = form.model.data
 
+        # print("Image id: {}".format(image_id))
```
Comment (maurapintor) [inline app/routes/classifications.py]: Please, remove unnecessary print statements and comments. They have probably been added for debugging purposes, right?

File: app/routes/downloadJson.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                        headers={"Content-disposition":
+                        "attachment; filename={}.json".format(jobId)})
+        
+        
+
+    
+    elif request.method == 'GET':
+        return '''
+        <html><body>
+        <h1> omage </h1>
```
Comment (maurapintor) [inline app/routes/downloadJson.py]: typo?

File: app/routes/downloadPlotPng.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        
+        return send_from_directory(stripped, filename,as_attachment=True)
+        
+        
+
+    
+    elif request.method == 'GET':
+        return '''
+        <html><body>
+        <h1> omage </h1>
```
Comment (maurapintor) [inline app/routes/downloadPlotPng.py]: typo again?

**Review by maurapintor — CHANGES_REQUESTED** [review]
Overall, the implementation is already good 👍 . 
Some things can still be improved before I merge this to the main repository:
* there is a `.DS_Store` file inside the `app` folder that was probably pushed by mistake.
* I left some comments in the code because there are several commented lines and debug prints that should be removed.
* There are some typos in the HTML (probably, not sure whether they were intended or not).

File: app/routes/downloadPlotPng.py (hunk 4 above)
Comment (AntonioCampus) [reply app/routes/downloadPlotPng.py]: Ok i cleaned the code

**Review by maurapintor — APPROVED** [review]
Looks good to me!
Closing the PR and merging the contribution.
Well done! 👍

**Pull request merged**